models/repo/setting.go

```diff
@@ -0,0 +1,231 @@
+	for _, s := range settings {
+		settingsMap[s.SettingKey] = s
+	}
+	return settingsMap, nil
+}
+
+// DeleteSetting deletes a specific setting for a repo
+func DeleteSetting(repoID int64, key string) error {
+	if err := validateRepoSettingKey(key); err != nil {
+		return err
+	}
+	cache.Remove(genSettingCacheKey(repoID, key))
+	_, err := db.GetEngine(db.DefaultContext).Delete(&Setting{RepoID: repoID, SettingKey: key})
+	return err
+}
+
+func SetSettingNoVersion(repoID int64, key, value string) error {
+	s, err := GetSettingNoCache(repoID, key)
+	if errors.Is(err, util.ErrNotExist) {
+		return SetSetting(&Setting{
+			RepoID:       repoID,
+			SettingKey:   key,
+			SettingValue: value,
+		})
+	}
+	if err != nil {
+		return err
+	}
+	s.SettingValue = value
+	return SetSetting(s)
+}
+
+// SetSetting updates a users' setting for a specific key
+func SetSetting(setting *Setting) error {
+	if err := upsertSettingValue(setting.RepoID, strings.ToLower(setting.SettingKey), setting.SettingValue, setting.Version); err != nil {
+		return err
+	}
+
+	setting.Version++
+
+	cc := cache.GetCache()
+	if cc != nil {
+		return cc.Put(genSettingCacheKey(setting.RepoID, setting.SettingKey), setting.SettingValue, setting_module.CacheService.TTLSeconds())
+	}
+
+	return nil
+}
+
+func upsertSettingValue(repoID int64, key, value string, version int) error {
+	return db.WithTx(db.DefaultContext, func(ctx context.Context) error {
+		e := db.GetEngine(ctx)
+
+		// here we use a general method to do a safe upsert for different databases (and most transaction levels)
+		// 1. try to UPDATE the record and acquire the transaction write lock
+		//    if UPDATE returns non-zero rows are changed, OK, the setting is saved correctly
+		//    if UPDATE returns "0 rows changed", two possibilities: (a) record doesn't exist  (b) value is not changed
+		// 2. do a SELECT to check if the row exists or not (we already have the transaction lock)
+		// 3. if the row doesn't exist, do an INSERT (we are still protected by the transaction lock, so it's safe)
+		//
+		// to optimize the SELECT in step 2, we can use an extra column like `revision=revision+1`
+		//    to make sure the UPDATE always returns a non-zero value for existing (unchanged) records.
+
+		res, err := e.Exec("UPDATE repo_setting SET setting_value=?, version = version+1 WHERE repo_id=? AND setting_key=? AND version=?", value, repoID, key, version)
+		if err != nil {
+			return err
+		}
+		rows, _ := res.RowsAffected()
+		if rows > 0 {
+			// the existing row is updated, so we can return
+			return nil
+		}
+
+		// in case the value isn't changed, update would return 0 rows changed, so we need this check
+		has, err := e.Exist(&Setting{RepoID: repoID, SettingKey: key})
+		if err != nil {
+			return err
+		}
+		if has {
+			return nil
+		}
+
+		// if no existing row, insert a new row
+		_, err = e.Insert(&Setting{RepoID: repoID, SettingKey: key, SettingValue: value})
+		return err
+	})
+}
```

Comment:
Hmm, I'm wondering whether we can extract all this code and unify it with the `user specific settings`?
So that we simply pass the table to operate on in the current methods, and do the rest in the common version.
The only problem might be `&Setting{RepoID: repoID,…}`.

---

There is also another possible approach:
Instead of defining a separate type, we could also just add `RepoID` to the existing user secrets.
 That would have the least overhead, and it would even allow for repo-specific user settings.

Reply:
> Hmm, I'm wondering whether we can extract all this code and unify it with the `user specific settings`? So that we simply pass the table to operate on in the current methods, and do the rest in the common version. The only problem might be `&Setting{RepoID: repoID,…}`.
> 
> There is also another possible approach: Instead of defining a separate type, we could also just add `RepoID` to the existing user secrets. That would have the least overhead, and it would even allow for repo-specific user settings. So,
> 
> * `UserID = 0` would mean `repo secret`,
> * `RepoID = 0` would mean `user secret`,
> * `UserID != 0 && RepoID != 0` would mean `user secret for a given repo`
> * `UserID = 0 && RepoID = 0` would be an interesting case: Is that invalid, or is that a system setting?

I don't think it's a good idea to unify the two tables. For a better performance, we should create more tables but not store all records in one table. Copy is better than complicated abstract when there is only small duplication. Come from Golang's principle if I'm not wrong.

Reply:
`Small duplication`? When you copy literally **everything** except changing `UserID` with `RepoID`?
My proposed approach has the benefits that it can store all four states, all of which might be wanted by someone…